[quant] Make PerChannel Observer work with float qparams

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #42881 [quant] Add torchbind support for embedding_bag packed weights
* #42762 [quant] Add embeddingbag_prepack function that works on quantized tensor.
* **#42690 [quant] Make PerChannel Observer work with float qparams**
* #42612 [quant] Create PerRowQuantizer for floating point scale and zero_point

Summary:
Add implementation for new qscheme per_channel_affine_float_qparams in observer

Test Plan:
python test/test_quantization.py TestObserver.test_per_channel_observers

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23070633](https://our.internmc.facebook.com/intern/diff/D23070633)

supriyar committed:
[quant] Make PerChannel Observer work with float qparams
Summary:
Add implementation for new qscheme per_channel_affine_float_qparams in observer

Test Plan:
python test/test_quantization.py TestObserver.test_per_channel_observers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented on torch/quantization/observer.py:
```diff
@@ -247,6 +254,9 @@ def _calculate_qparams(self, min_val, max_val):
             # TODO: switch to zero_point.item() after adding JIT support
             zero_point = torch.tensor([int(zero_point)], dtype=zero_point.dtype)
+            if self.qscheme == torch.per_channel_affine_float_qparams:
+                zero_point = torch.tensor([float(zero_point)], dtype=zero_point.dtype)
```
nit: should we also set the device?

vkuzo commented on torch/quantization/observer.py:
```diff
@@ -232,6 +234,11 @@ def _calculate_qparams(self, min_val, max_val):
+        elif self.qscheme == torch.per_channel_affine_float_qparams:
+            scale = (orig_max - orig_min) / float(qmax - qmin)
+            scale_ones = torch.ones_like(scale)
+            scale = torch.where(scale != 0, scale, scale_ones)
```
qq, should this be equality comparison or a threshold check against zero?

supriyar replied:
Do you mean check for `scale>0` vs `scale!=0`? In the code we compute inverse scale so we need to set all values of 0 scale to 1.

vkuzo replied:
`abs(scale) < eps` instead of `scale != 0`, since scale is a float...?

vkuzo commented on torch/quantization/observer.py:
```diff
@@ -214,7 +215,8 @@ def _calculate_qparams(self, min_val, max_val):
-
+        orig_min = min_val
+        orig_max = max_val
         min_val = torch.min(min_val, torch.zeros_like(min_val))
         max_val = torch.max(max_val, torch.zeros_like(max_val))
```
this code was a bit confusing (before this PR).  Maybe we can rename these something like `min_val_neg` and `max_val_pos` in the rest of the function?

vkuzo commented on torch/quantization/observer.py:
```diff
@@ -232,6 +234,11 @@ def _calculate_qparams(self, min_val, max_val):
                 else:
                     zero_point = zero_point.new_full(zero_point.size(), 128)
+        elif self.qscheme == torch.per_channel_affine_float_qparams:
+            scale = (orig_max - orig_min) / float(qmax - qmin)
```
ideally this should be `max_val - min_val`, since that's what is actually happening.  The other qschemes are not using observed min and max directly.

supriyar replied:
Right. Maybe I can rename the other usages so I can use `max_val - min_val` directly here

supriyar committed:
Update on "[quant] Make PerChannel Observer work with float qparams"


Summary:
Add implementation for new qscheme per_channel_affine_float_qparams in observer

Test Plan:
python test/test_quantization.py TestObserver.test_per_channel_observers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented on torch/quantization/observer.py:
```diff
@@ -214,39 +215,49 @@ def _calculate_qparams(self, min_val, max_val):
-        else:
+        elif self.qscheme == torch.per_channel_affine_float_qparams:
             scale = (max_val - min_val) / float(qmax - qmin)
+            scale = torch.where(scale > self.eps, scale, torch.ones_like(scale))
```
abs(scale)?

vkuzo replied:
sorry, never mind, ignore this

jerryzh168 commented on test/quantization/test_workflow_module.py:
```diff
@@ -281,11 +281,15 @@ def test_per_tensor_dynamic_quant_observers(self, X, reduce_range):
            ch_axis=st.sampled_from((0, 1, 2, 3)), reduce_range=st.booleans())
     def test_per_channel_observers(self, qdtype, qscheme, ch_axis, reduce_range):
         # reduce_range cannot be true for symmetric quantization with uint8
+        qscheme = torch.per_channel_affine_float_qparams
```
don't forget to review this before landing

supriyar committed:
Update on "[quant] Make PerChannel Observer work with float qparams"


Summary:
Add implementation for new qscheme per_channel_affine_float_qparams in observer

Test Plan:
python test/test_quantization.py TestObserver.test_per_channel_observers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Make PerChannel Observer work with float qparams"


Summary:
Add implementation for new qscheme per_channel_affine_float_qparams in observer

Test Plan:
python test/test_quantization.py TestObserver.test_per_channel_observers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Make PerChannel Observer work with float qparams"


Summary:
Add implementation for new qscheme per_channel_affine_float_qparams in observer

Test Plan:
python test/test_quantization.py TestObserver.test_per_channel_observers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Make PerChannel Observer work with float qparams"


Summary:
Add implementation for new qscheme per_channel_affine_float_qparams in observer

Test Plan:
python test/test_quantization.py TestObserver.test_per_channel_observers

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23070633](https://our.internmc.facebook.com/intern/diff/D23070633)

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Make PerChannel Observer work with float qparams"


Summary:
Add implementation for new qscheme per_channel_affine_float_qparams in observer

Test Plan:
python test/test_quantization.py TestObserver.test_per_channel_observers

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23070633](https://our.internmc.facebook.com/intern/diff/D23070633)

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Make PerChannel Observer work with float qparams"


Summary:
Add implementation for new qscheme per_channel_affine_float_qparams in observer

Test Plan:
python test/test_quantization.py TestObserver.test_per_channel_observers

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23070633](https://our.internmc.facebook.com/intern/diff/D23070633)

[ghstack-poisoned]
